pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go

```diff
@@ -38,48 +57,89 @@ var (
 
 // NewMustMatchPatterns creates a new mustMatchPatterns strategy that will provide validation.
 // Passing nil means the default pattern, passing an empty list means to disallow all sysctls.
-func NewMustMatchPatterns(patterns []string) SysctlsStrategy {
-	if patterns == nil {
-		patterns = defaultSysctlsPatterns
-	}
+func NewMustMatchPatterns(safeWhitelist, allowedUnsafeSysctls, forbiddenSysctls []string) SysctlsStrategy {
 	return &mustMatchPatterns{
-		patterns: patterns,
+		safeWhitelist:        safeWhitelist,
+		allowedUnsafeSysctls: allowedUnsafeSysctls,
+		forbiddenSysctls:     forbiddenSysctls,
 	}
 }
 
-// Validate ensures that the specified values fall within the range of the strategy.
-func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
-	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.SysctlsPodAnnotationKey)...)
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.UnsafeSysctlsPodAnnotationKey)...)
-	return allErrs
+func (s *mustMatchPatterns) divideSysctls(sysctls []api.Sysctl) (safeSysctls, unsafeSysctls []sysctlItem) {
+	for i, sysctl := range sysctls {
+		found := false
+		for _, ws := range s.safeWhitelist {
+			if sysctl.Name == ws {
+				found = true
+				break
+			}
+		}
+		if found {
+			safeSysctls = append(safeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		} else {
+			unsafeSysctls = append(unsafeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		}
+	}
+	return
 }
 
-func (s *mustMatchPatterns) validateAnnotation(pod *api.Pod, key string) field.ErrorList {
+// Validate ensures that the specified values fall within the range of the strategy.
+func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	fieldPath := field.NewPath("pod", "metadata", "annotations").Key(key)
-
-	sysctls, err := helper.SysctlsFromPodAnnotation(pod.Annotations[key])
-	if err != nil {
-		allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], err.Error()))
+	var sysctls []api.Sysctl
+	if pod.Spec.SecurityContext != nil {
+		sysctls = pod.Spec.SecurityContext.Sysctls
 	}
 
 	if len(sysctls) > 0 {
-		if len(s.patterns) == 0 {
-			allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], "sysctls are not allowed"))
-		} else {
-			for i, sysctl := range sysctls {
-				allErrs = append(allErrs, s.ValidateSysctl(sysctl.Name, fieldPath.Index(i))...)
+		fieldPath := field.NewPath("pod", "spec", "securityContext")
+
+		safeSysctls, unsafeSysctls := s.divideSysctls(sysctls)
```

Comment:
I expected something more like this:
```
for i, sysctl := range sysctls {
  switch {
    case s.isForbidden(sysctl.Name):
      ... add forbidden error ...
    case s.isSafe(sysctl.Name):
      continue
    case s.isAllowedUnsafe(sysctl.Name):
      continue
    default:
      ... add "not allowed unsafe" error ...
  }
}
```

that does several things:
* keeps us from having to divide the sysctls
* keeps the helper functions really small and testable (input is a name, output is a boolean)
* avoids plumbing indexes everywhere
* ensures every sysctl gets run through the same logic (instead of operating on parallel lists differently)

Reply:
We have a semantical problem here:
- assume in 1.10 the sysctl `foo.bar` is unsafe and whitelisted (in the PSP and the kubelet). At the same time `forbiddenSysctls: ["foo.*"]`.
- now in 1.11 `foo.bar` is declared to become safe. It suddenly is forbidden.

Reply:
^^ what we need: a sysctl explicitly allowed as unsafe sysctls will pass the admission even though it matches a pattern in the `forbiddenSysctls` list. In other words: being in the `allowedUnsafeSysctls` list trumps over the forbidden list, independencly whether the actual sysctl is safe or not.

Reply:
No, forbidden should not be allowed to cover anything in the allowedUnsafeSysctls list

Reply:
> foo.bar is unsafe and whitelisted (in the PSP and the kubelet). At the same time forbiddenSysctls: ["foo.*"].

Validation should prevent anything in forbiddenSysctls from covering anything in allowedUnsafeSysctls

Reply:
we would loose the case `*` + `some-unsafe-sysctl`, which might be common

Reply:
`forbiddenSafeSysctls` + `whitelistedSysctls` (not the missing "unsafe") would logically solve this

Reply:
Update to
```go
for i, sysctl := range sysctls {
  switch {
    case s.isForbidden(sysctl.Name):
      ... add forbidden error ...
    case s.isSafe(sysctl.Name):
      continue
    case s.isAllowedUnsafe(sysctl.Name):
      continue
    default:
      ... add "not allowed unsafe" error ...
  }
}
```